Texture manager changes

Freezarite opened the pull request.

Freezarite committed:
made texture_cache map private

Freezarite committed:
changed texture_cache map to use enums instead of string for keys and functions using said map

Nanogamer7 commented on docs/doxygen-awesome-css:
whad this do?

Nanogamer7 commented on include/SpriteComponent.h:
```diff
@@ -4,10 +4,12 @@
 #include "AnimationHandler.h"
 #include "Component.h"
 #include "Direction.h"
+#include "TextureEnumBase.h"
```
duplicate include

Nanogamer7 commented on include/TextureEnumBase.h:
```diff
@@ -0,0 +1,3 @@
+#pragma once
+
+enum class TexturesEnum;
```
Missing documentation

Nanogamer7 commented on include/TextureManager.h:
```diff
@@ -17,11 +18,12 @@ class TextureManager
-
-		SDL_Texture* loadTexture(const char* fileName);
+		void addSingleTexture(TexturesEnum texture, const char* filePath);
+		void addTextures(const std::map<TexturesEnum, const char*>& textures);
```
Missing documentation

Nanogamer7 commented on include/TileComponent.h:
```diff
@@ -17,17 +19,27 @@ class TileComponent : public Component
+		return std::string(magic_enum::enum_name(this->texture));
+#else
+		return "Undefined Enum";
+#endif
```
Compiler declarations make the code less readable, avoid if possible

Nanogamer7 replied:
This can probably be removed, as either a compiler error will be thrown, or one can be thrown using `static_assert`. Either way, please remove compiler declaration

Nanogamer7 replied:
As discussed, this function isn't used, thus can be removed. This also means `magic_enum` is not needed.

Nanogamer7 commented on src/AssetManager.cpp:
```diff
@@ -15,12 +15,15 @@
 
 AssetManager::~AssetManager() {}
 
+//seems to not be used anymore
```
Indeed not used anymore, can be deleted

Nanogamer7 commented on src/GameInternal.cpp:
```diff
@@ -78,10 +78,11 @@ void GameInternal::init(const char* title, int xpos, int ypos, int width, int he
 
-	SDL_Texture* backgroundTexture = GameInternal::textureManager->loadTexture("assets/startscreen.png");
+
+	//SDL_Texture* backgroundTexture = GameInternal::textureManager->loadTexture("assets/startscreen.png");
```
Should be reimplemented in chickengame

Nanogamer7 commented on src/TextureManager.cpp:
```diff
@@ -6,17 +6,32 @@
-		return it->second;
+
+void TextureManager::addSingleTexture(TexturesEnum texture, const char* filePath) {
+	auto sdlTexture = IMG_LoadTexture(this->manager->getGame()->renderer, filePath);
```
On merge: replace `this->manager->getGame()` with `VEGO_Game()`

Nanogamer7 commented on src/TextureManager.cpp:
```diff
@@ -6,17 +6,32 @@
-	if (it != this->texture_cache.end()) {
-		return it->second;
+
+void TextureManager::addSingleTexture(TexturesEnum texture, const char* filePath) {
```
This is problematics as it doesn't allow adding automatically provided textures (i.e. map tilesheets) - which would require a generic addTexture method without an identifier

Nanogamer7 reviewed (changes requested):
`textureEnum` is a redundant name, as enum is already the type, just name the variables `texture` or `textureId` instead

Freezarite committed:
magic_enum library stuff and refactored TextureEnum to Textures due to bad naming

Freezarite committed:
merged dev into textureManagerChanges

Freezarite committed:
added new Map for tile-textures as they wont work with our current enum-class maps.

Freezarite committed:
added new Constructor for SpriteComponent as Tiles wont work with enums

Freezarite committed:
Extra map for MapTiles generated by TMX file now works like intended.
removed textures map from AssetManager as it is no longer used.
Updated SpriteComponent to now check if it is a normal Texture or a MapTileTexture.
Added if condition in TextureManager::LoadMapTileTexture to check if the texture was able to be loaded by SDL

Freezarite committed:
Added some doxygen documentation to TextureEnumBase.h and TextureManager.h

Nanogamer7 commented:
Force pushes due to faulty/unnecessary merge

Nanogamer7 commented on include/SpriteComponent.h:
```diff
@@ -34,18 +35,23 @@ class SpriteComponent : public Component, public RenderObject
 	int textureYOffset;
 
+	//should be changed in the future as this is only for the tiles
+	const char* path;
```
add more descriptive comment

Nanogamer7 commented on include/TextureEnumBase.h:
```diff
@@ -0,0 +1,18 @@
+ * \brief Forward declaration of the \c Textures enum class.
+ *
+ * This file contains a forward declaration of the \c Textures enum class, which is used as a base
+ * class for texture identifiers in the engine. It allows developers to define their own set of texture
```
Reword to not mention developers directly, as this sounds like internal documentation

Nanogamer7 commented on include/TextureEnumBase.h:
```diff
@@ -0,0 +1,18 @@
+ * class for texture identifiers in the engine. It allows developers to define their own set of texture
+ * types in their own implementation, providing flexibility in texture management within the engine.
+ *
+ * \details
```
Above untagged text and details should not both exist

Nanogamer7 commented on include/TextureManager.h:
```diff
@@ -6,6 +6,23 @@
+ * map tile textures (`mapTile_texture_cache`). This allows for efficient texture reuse
+ * and management across different parts of the game.
+ *
+ * \note Textures are identified by an enum class `Textures` which is user-defined.
```
link to texture enum using [`\sa`](https://www.doxygen.nl/manual/commands.html#cmdsa)

Nanogamer7 commented on include/TextureManager.h:
```diff
@@ -6,6 +6,23 @@
+ * and management across different parts of the game.
+ *
+ * \note Textures are identified by an enum class `Textures` which is user-defined.
+ *       Developers are expected to define the texture types in their own implementation.
```
replace "developer" with "the game" or similar

Nanogamer7 commented on include/TextureEnumBase.h:
```diff
@@ -0,0 +1,18 @@
+/*!
+ * \file Textures.h
```
This does not generate any documentation, as the file `Textures.h` does not exist. Refer to the [doxgen documentation](https://www.doxygen.nl/manual/commands.html#cmdfile) for usage. Either way, the documentation should be on the enum class, not the file itself (i.e. after `#pragma once`)

Nanogamer7 commented on include/TextureManager.h:
Names of private members do not concern the developer, or are even visible for them, and such referencing them only causes confusion and extra complexity. Please rewrite by only describing functionality on a higher level, which concerns the game developer.

(`mapTile_texture_cache`, `texture_cache`, etc.)

Nanogamer7 commented on include/TextureEnumBase.h:
```diff
@@ -0,0 +1,18 @@
+
+#pragma once
+
+enum class Textures;
```
File and class name should be equal

Nanogamer7 commented on include/TextureManager.h:
```diff
@@ -15,13 +32,59 @@ class TextureManager
+		 * \param path The file path to the texture.
+		 * \return A pointer to the `SDL_Texture` representing the map tile.
+		 *
+		 * This function checks if the map tile texture is already cached. If not, it
```
Caching does not concern the developer, it can be mentioned at the documentation for the class itself, but does not add any information to the usage of this function

Nanogamer7 commented on include/TextureManager.h:
```diff
@@ -15,13 +32,59 @@ class TextureManager
+		 * This function checks if the map tile texture is already cached. If not, it
+		 * loads the texture from the file system and stores it in the cache.
+		 *
+		 * If loading fails, an exception is thrown with a descriptive error message.
```
What exception is thrown? required info for catching/handling. "Descriptive error message" is expected and thus unnecessary.

Freezarite committed:
Cleanup of documentation and refactoring
renamed TextureEnumBase.h to Textures.h
improved some of the doxygen documentation

Freezarite committed:
fixed \todo in TextureManager.h

Nanogamer7 commented on include/SpriteComponent.h:
```diff
@@ -35,8 +35,9 @@ class SpriteComponent : public Component, public RenderObject
-	const char* path;
+	//might be a better solution as this variable is only used for the loading of the tmx map
+	//TODO: improve this in the future
+	const char* path; //empty string if texture has a texture enum value, otherwise the path of the texture
```
Should probably be a doxygen comment

Nanogamer7 commented on include/SpriteComponent.h:
```diff
@@ -35,8 +35,9 @@ class SpriteComponent : public Component, public RenderObject
 
-	//should be changed in the future as this is only for the tiles
-	const char* path;
+	//might be a better solution as this variable is only used for the loading of the tmx map
```
what is this referring too?
"might be" - what might be?

Nanogamer7 commented on include/TextureManager.h:
```diff
@@ -16,12 +16,8 @@
- *
- * \note Textures are identified by an enum class `Textures` which is user-defined.
- *       Developers are expected to define the texture types in their own implementation.
+ * \sa \ref Textures "Textures" are used to identify textures within the engine.
```
`\ref` is not required, see https://www.doxygen.nl/manual/autolink.html

Freezarite committed:
removed magic enum and documentation improvements

since getName is no longer used from the TileComponent.h we no longer need magic_enum for the enum to string conversion
also some minor documentation changes

Freezarite committed:
more documentation changes

Nanogamer7 reviewed (approved):
lgtm
